index-providers/modeshape-elasticsearch-index-provider/src/main/java/org/modeshape/jcr/index/elasticsearch/EsIndexProvider.java

```diff
@@ -0,0 +1,197 @@
+/*
+ * ModeShape (http://www.modeshape.org)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.modeshape.jcr.index.elasticsearch;
+
+import java.util.Collection;
+import javax.jcr.RepositoryException;
+import javax.jcr.query.qom.ChildNodeJoinCondition;
+import javax.jcr.query.qom.DescendantNodeJoinCondition;
+import javax.jcr.query.qom.DynamicOperand;
+import javax.jcr.query.qom.JoinCondition;
+import org.elasticsearch.action.admin.cluster.health.ClusterHealthResponse;
+import org.elasticsearch.action.admin.cluster.health.ClusterHealthStatus;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.settings.ImmutableSettings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.node.Node;
+import org.elasticsearch.node.NodeBuilder;
+import org.modeshape.common.collection.Problems;
+import org.modeshape.jcr.ExecutionContext;
+import org.modeshape.jcr.NodeTypes;
+import org.modeshape.jcr.api.index.IndexDefinition;
+import org.modeshape.jcr.api.query.qom.ChildCount;
+import org.modeshape.jcr.cache.change.ChangeSetAdapter;
+import org.modeshape.jcr.query.QueryContext;
+import org.modeshape.jcr.query.model.FullTextSearch;
+import org.modeshape.jcr.query.model.Or;
+import org.modeshape.jcr.spi.index.IndexCostCalculator;
+import org.modeshape.jcr.spi.index.provider.IndexProvider;
+import org.modeshape.jcr.spi.index.provider.IndexUsage;
+import org.modeshape.jcr.spi.index.provider.ManagedIndexBuilder;
+
+/**
+ *
+ * @author kulikov
+ */
+public class EsIndexProvider extends IndexProvider {
```

Comment:
this class should also override the ` protected int getCostEstimate()` method from the base class, making sure it returns a `REMOTE` cost. In other words, if there's a local index (MapDB or Lucene) that covers the same property, it should be preferred over the `REMOTE` one.